### PR TITLE
[fix] Fix nested FlattenParamsWrapper state_dict/load_state_dict

### PR DESCRIPTION
**Pull request by myleott**

### CODE REVIEW

File: fairscale/nn/misc/flatten_params_wrapper.py [hunk 1]
```diff
@@ -131,21 +162,53 @@ def __getattr__(self, name: str) -> Any:
...
 
     def flat_state_dict(self, *args: Any, **kwargs: Any) -> Dict[str, Any]:
         """Return the flattened state_dict."""
-        return super().state_dict(*args, **kwargs)
+        assert self.is_flattened
+        with ExitStack() as stack:
+            # tell any nested FlattenParamsWrapper instances not to auto unflatten
+            for module in self.modules():
+                if isinstance(module, FlattenParamsWrapper):
+                    stack.enter_context(module._no_auto_unflatten_state_dict())
```
Comment (sshleifer) [inline fairscale/nn/misc/flatten_params_wrapper.py:179]: Very nice

**Comment by min-xu-ai** [comment]
Thanking out loud, instead of trying to hack around all the recursive thing in state_dict and load_state_dict, can we can operate on the resulting dict directly?

state_dict (vanilla version, no FSDP) -> some preprocessing code -> state_dict (compatible with FSDP/flatten_param) -> hand it to load_state_dict() of pytorch.

The trick is the "some preprocessing code". I'd image that code operates on pure dict data, perhaps even remove unneeded shards to reduce its size. But would that code be easier/harder to write?

**Comment by myleott** [comment]
Yeah, I think we probably should go down that route. One idea is that instead of storing the wrapped stuff in `self.module`, we could give it a more grep'able name, like `self._fsdp_wrapped_module`, `self._fpw_wrapped_module`, etc.

Alternatively we could try to build an API on top of state_dicts that looks more like this: tinyurl.com/32xm4ccb

**Comment by sshleifer** [comment]
**TLDR:** happy to help, not sure which option is best.

**Thinking out loud:**


0) I think this PR is a good change.

1) I feel 7/10 strongly that any approach to the full problem is significantly better than letting the user figure it out. It's frustrating to print out two state dicts side by side, know that they are basically the same, know that someone else has printed out the exact same thing before and hacked around it and moved on, then hack around it and move on.

2) This feeling, and warm memories of hacking many a state dict for hf/transformers conversions, makes me think that min's proposal above may be the way to go for `FSDP`, especially if we can keep pretending that `FlattenParametersWrapper` doesn't know about `FSDP` but `FSDP` knows about `FPW`.

3)  Maybe a pipe dream, but this all gets a lot simpler if we can remove `flatten_parameters=True/False` optionality in `FSDP`. IIRC, @min-xu-ai has a very good reason to use `flatten_parameters=False`?

**Comment by min-xu-ai** [comment]
> IIRC, @min-xu-ai has a very good reason to use flatten_parameters=False?

Don't know it was a "good reason" but I was forced to use it because BatchNorm wants to be special in mixed precision. It needs to run with fp32 states, so we can't have it flattened with other fp16 params. :-( Don't really know how common this really it in other workloads. Even with vision, they are many papers talking about training without BN. Unfortunately, presently VISSL still need BN as well as SyncBN (for DDP).

**Comment by myleott** [comment]
Okay, lemme take a stab at something clean here.

Re: 3, we're stuck with flatten_params for now, due to huge performance drop without it (see benchmarks here: https://github.com/facebookresearch/fairscale/issues/430#issuecomment-785463057). After bucketing is fully implemented, it should mostly close the gap, but not fully.

Re: BatchNorm, we should come up with a solution here. Perhaps we can keep separate flattened states for some modules, not sure. We should add some tests that illustrate the BN problem though...

**Comment by min-xu-ai** [comment]
>  We should add some tests that illustrate the BN problem though...
Yeah, this is on my todo.

**Comment by min-xu-ai** [comment]
Did you see the state_dict hook?

https://github.com/pytorch/pytorch/blob/569d4fe3f95af913d5cc20ded09c2d000e974cc9/torch/nn/modules/module.py#L1081

Maybe it can be used to generate necessary info during state_dict generation and then used in load_state_dict?

**Comment by myleott** [comment]
Nice find! `_register_load_state_dict_pre_hook` seems perfect. Each nested FSDP/FlattenParamsWrapper can modify the state dict right before loading

File: fairscale/nn/misc/flatten_params_wrapper.py [hunk 2]
```diff
@@ -131,22 +171,61 @@ def __getattr__(self, name: str) -> Any:
...
-        with self.unflatten_params():
-            return self.module.state_dict(*args, **kwargs)
+        """Return the wrapped module's state_dict (unflattened)."""
+        if self.is_flattened and self._auto_unflatten_state_dict:
+            with self.unflatten_params(recurse=False):
+                return super().state_dict(*args, **kwargs)
+        else:
+            return super().state_dict(*args, **kwargs)
+
+    def _post_state_dict_hook(
```
Comment (min-xu-ai) [inline fairscale/nn/misc/flatten_params_wrapper.py:181]: make this a static method too? In fact, why not make both hooks normal methods outside of the class? They don't need "self", right?

File: fairscale/nn/misc/flatten_params_wrapper.py [hunk 3]
```diff
@@ -110,18 +133,35 @@ def _unflatten_params(self) -> None:
...
+        If the current instance is already unflattened, then it will remain
+        unflattened after the context manager exits.
+        """
+        if recurse:
+            with ExitStack() as stack:
+                for module in self.modules():
+                    if isinstance(module, FlattenParamsWrapper):
+                        stack.enter_context(module.unflatten_params(recurse=False))
+                yield
+            return
```
Comment (min-xu-ai) [inline fairscale/nn/misc/flatten_params_wrapper.py:156]: how about change this to a "else:"? early return can cause future common code to be skipped by mistaken.

File: fairscale/nn/misc/flatten_params_wrapper.py [hunk 4]
```diff
@@ -110,18 +133,35 @@ def _unflatten_params(self) -> None:
...
-        self._unflatten_params()
+    def unflatten_params(self, recurse: bool = True) -> Generator:
+        """
+        Unflatten params (optionally recursively on all nested instances).
+        If the current instance is already unflattened, then it will remain
+        unflattened after the context manager exits.
+        """
+        if recurse:
+            with ExitStack() as stack:
+                for module in self.modules():
```
Comment (min-xu-ai) [inline fairscale/nn/misc/flatten_params_wrapper.py:152]: add a bit comment on this loop? It is not common to see loop and enter context in a context mgr. It is unclear what's the semantics to the caller.

File: fairscale/nn/misc/flatten_params_wrapper.py [hunk 5]
```diff
@@ -131,22 +171,61 @@ def __getattr__(self, name: str) -> Any:
             return getattr(self.module, name)  # fallback to wrapped module
 
     def state_dict(self, *args: Any, **kwargs: Any) -> "OrderedDict[str, Tensor]":  # type: ignore
-        """Return an unflattened state_dict."""
-        with self.unflatten_params():
-            return self.module.state_dict(*args, **kwargs)
+        """Return the wrapped module's state_dict (unflattened)."""
+        if self.is_flattened and self._auto_unflatten_state_dict:
+            with self.unflatten_params(recurse=False):
```
Comment (min-xu-ai) [inline fairscale/nn/misc/flatten_params_wrapper.py:176]: nice!

File: fairscale/nn/misc/flatten_params_wrapper.py [hunk 6]
```diff
@@ -131,22 +171,61 @@ def __getattr__(self, name: str) -> Any:
...
+            state_dict = self.state_dict(*args, **kwargs)
+        return state_dict
+
+    @contextmanager
+    def _no_auto_unflatten_state_dict(self) -> Generator:
+        backup = self._auto_unflatten_state_dict
+        self._auto_unflatten_state_dict = False
+        yield
+        self._auto_unflatten_state_dict = backup
+        return
```
Comment (min-xu-ai) [inline fairscale/nn/misc/flatten_params_wrapper.py:204]: extra return?

File: fairscale/utils/state_dict.py [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+    """
+    Replace all keys that match a given old_prefix with a new_prefix (in-place).
+
+    Usage::
+
+        state_dict = {"layer.xyz": torch.tensor(1)}
+        replace_by_prefix_(state_dict, "layer.", "module.layer.")
+        assert state_dict == {"module.layer.xyz": torch.tensor(1)}
+    """
+    for key in list(state_dict.keys()):
```
Comment (min-xu-ai) [inline fairscale/utils/state_dict.py:68]: assert old_prefix != new_prefix? If they equal, I think we will loss items.

**Review by min-xu-ai — APPROVED** [review]
very cool!

**Review by sshleifer — APPROVED** [review]

File: stubs/torch/__init__.pyi [hunk 8]
```diff
@@ -28,6 +28,7 @@ from . import cuda as cuda
 from . import optim as optim
 from . import nn as nn
 from . import testing as testing
+from . import utils as utils
```
Comment (min-xu-ai) [inline stubs/torch/__init__.pyi:31]: I am including these two unrelated changed in a separate PR. If it goes in first, you can remove these. otherwise, merge them in, I will deal with conflicts.
